app/preload/database/accountDao.ts

```diff
@@ -98,6 +103,93 @@ export async function updateInitialAccount(
         .update(updatedValues);
 }
 
+/** Inserts the given account as part of a transaction
+ * Also inserts a addressbookentry for the account, if it does not already exist.
+ */
+async function insertAccountTransactionally(
+    account: Account,
+    note: string,
+    transaction: Knex.Transaction
+) {
+    const abe = await transaction
+        .table(addressBookTable)
+        .where({ address: account.address })
+        .first()
+        .select();
+    if (abe) {
+        await transaction
+            .table(accountsTable)
+            .insert({ ...account, name: abe.name });
+        await transaction
+            .table(addressBookTable)
+            .where({ address: account.address })
+            .update({ readOnly: true });
+    } else {
+        await transaction.table(accountsTable).insert(account);
+        await transaction.table(addressBookTable).insert({
+            address: account.address,
+            name: account.name,
+            readOnly: true,
+            note,
+        });
+    }
+}
+
+/** Inserts accounts and credentials for a specific identity, from recovery.
+ * The identity is first inserted, and its given id is attached to the accounts and credentials.
+ */
+async function insertFromRecoveryNewIdentity(
+    recovered: AccountAndCredentialPairs,
+    identity: Omit<Identity, 'id'>
+) {
+    return (await knex()).transaction(async (transaction) => {
+        const identityId = (
+            await transaction.table(identitiesTable).insert(identity)
+        )[0];
+        for (const pair of recovered) {
+            const account = { ...pair.account, identityId };
+            const credential = { ...pair.credential, identityId };
+            const { address } = account;
+            const accountExists =
+                (
+                    await transaction
+                        .table(accountsTable)
+                        .where({ address })
+                        .select()
+                ).length > 0;
+            if (!accountExists) {
+                await insertAccountTransactionally(
+                    account,
+                    'Recovered account',
+                    transaction
+                );
+            }
+            await transaction.table(credentialsTable).insert(credential);
+        }
+    });
+}
+
+/** Inserts accounts and credentials for an existing identity, from recovery.
+ */
+async function insertFromRecoveryExistingIdentity(
+    recovered: AccountAndCredentialPairs
+) {
+    return (await knex()).transaction(async (transaction) => {
+        for (const { account, credential } of recovered) {
+            const { address } = account;
+            const accountExists = (await findAccounts({ address })).length > 0;
```

Comment:
Instead of implementing a check for whether an account exists multiple times, please add a method for it as something like:
```
async function ... (address: string) {
    return getAccount(address) !== undefined;
}
```

Reply:
I've refactored it, so I only do it once. We can't use the non-transactional functions here though.